xds/src/test/java/io/grpc/xds/OrcaServiceImplTest.java

```diff
@@ -0,0 +1,296 @@
+    call.start(listener, new Metadata());
+    call.sendMessage(OrcaLoadReportRequest.newBuilder().build());
+    call.halfClose();
+    call.request(1);
+    OrcaLoadReport expect = OrcaLoadReport.newBuilder().putUtilization("omg", 100).build();
+    verify(listener).onMessage(eq(expect));
+    defaultTestService.setMemoryUtilizationMetric(0.5);
+    ClientCall<OrcaLoadReportRequest, OrcaLoadReport> call2 = channel.newCall(
+        OpenRcaServiceGrpc.getStreamCoreMetricsMethod(), CallOptions.DEFAULT);
+    call2.start(listener, new Metadata());
+    call2.sendMessage(OrcaLoadReportRequest.newBuilder().build());
+    call2.halfClose();
+    call2.request(1);
+    expect = OrcaLoadReport.newBuilder(expect).setMemUtilization(0.5).build();
+    verify(listener).onMessage(eq(expect));
+    assertThat(defaultTestService.getClientsCount()).isEqualTo(2);
+    assertThat(fakeClock.getPendingTasks().size()).isEqualTo(2);
+    channel.shutdownNow();
+    assertThat(fakeClock.forwardTime(1, TimeUnit.SECONDS)).isEqualTo(0);
+    assertThat(defaultTestService.getClientsCount()).isEqualTo(0);
+    ArgumentCaptor<Status> callCloseCaptor = ArgumentCaptor.forClass(null);
+    verify(listener, times(2)).onClose(callCloseCaptor.capture(), any());
+    assertThat(callCloseCaptor.getValue().getCode()).isEqualTo(Status.Code.UNAVAILABLE);
+  }
+
+  @Test
+  public void testApis() throws Exception {
+    Map<String, Double> firstUtilization = ImmutableMap.of("util", 0.1);
+    OrcaLoadReport goldenReport = OrcaLoadReport.newBuilder()
+        .setCpuUtilization(random.nextDouble())
+        .setMemUtilization(random.nextDouble())
+        .putAllUtilization(firstUtilization)
+        .putUtilization("queue", 1.0)
+        .build();
+    defaultTestService.setCpuUtilizationMetric(goldenReport.getCpuUtilization());
+    defaultTestService.setMemoryUtilizationMetric(goldenReport.getMemUtilization());
+    defaultTestService.setAllUtilizationMetrics(firstUtilization);
+    defaultTestService.setUtilizationMetric("queue", 1.0);
+    Iterator<OrcaLoadReport> reports = OpenRcaServiceGrpc.newBlockingStub(channel)
+        .streamCoreMetrics(OrcaLoadReportRequest.newBuilder().build());
+    assertThat(reports.next()).isEqualTo(goldenReport);
+
+    defaultTestService.deleteCpuUtilizationMetric();
+    defaultTestService.deleteMemoryUtilizationMetric();
+    fakeClock.forwardTime(1, TimeUnit.SECONDS);
+    goldenReport = OrcaLoadReport.newBuilder()
+        .putAllUtilization(firstUtilization)
+        .putUtilization("queue", 1.0)
+        .putUtilization("util", 0.1)
+        .build();
+    assertThat(reports.next()).isEqualTo(goldenReport);
+    defaultTestService.deleteUtilizationMetric("util-not-exist");
+    defaultTestService.deleteUtilizationMetric("queue-not-exist");
+    fakeClock.forwardTime(1, TimeUnit.SECONDS);
+    assertThat(reports.next()).isEqualTo(goldenReport);
+
+    CountDownLatch latch = new CountDownLatch(1);
```

Comment:
a little bit of a concurrency test showing that it works when user application is updating metrics from another thread? probably not interesting but does not hurt.

Reply:
It doesn't actually test that though, because of the synchronization provided by the latch. Even if the code was not thread-safe this test would pass even with TSAN enabled.

Reply:
I guess so, because there is only one thread accessing the data since the new thread runs. 
How about simply repeating L271-L273 in the main thread, which would create the data race situation.

Reply:
You need to synchronize their start as well. `CyclicBarrier` is well suited to the task, for both before and after.

nit: instead of using an executor, you might consider `new Thread(new Runnable()...).start()` which won't leave the thread around after the test.